### PR TITLE
Span and Wagner EOS for CO2

### PR DESCRIPTION
**Pull request by cpgr**
This implements the Span and Wagner EOS for CO2 in the `fluid_properties` module. This is one of the main fluids that we use in our work, and this formulation (Span and Wagner) is considered the 'gold standard'.

Adds tests of all properties.

Note: This EOS requires iteration to find density given pressure and temperature. We have typically used Brent's method to find this root, so I have implemented a version in this PR. 

pinging @andrsd - I've only got one more fluid to add after this one so I'll stop making you find all of my formatting errors soon
Refs #6972

### CODE REVIEW

File: modules/fluid_properties/include/userobjects/CO2FluidProperties.h [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+ * Span and Wagner, "A New Equation of State for Carbon Dioxide Covering the Fluid Region
+ * from the Triple-Point Temperature to 1100K at Pressures up to 800 MPa",
+ * J. Phys. Chem. Ref. Data, 25 (1996)
+ *
+ * Note: the Span and Wagner EOS uses density and temperature as the primary variables. As
+ * a result, density must first be found using iteration, after which the other properties
+ * can be calculated directly.
+ */
+class CO2FluidProperties : public SinglePhaseFluidPropertiesPT
+
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/CO2FluidProperties.h]: No empty line

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+  for (unsigned int i = 0; i < _n2.size(); ++i)
+    phir += _n2[i] * std::pow(delta, _d2[i]) * std::pow(tau, _t2[i]) *
+      std::exp(- std::pow(delta, _c2[i]));
+
+  for (unsigned int i = 0; i < _n3.size(); ++i)
+    phir += _n3[i] * std::pow(delta, _d3[i]) * std::pow(tau, _t3[i]) *
+    std::exp(- _alpha3[i] * std::pow(delta - _eps3[i], 2.0) - _beta3[i] *
+    std::pow(tau - _gamma3[i], 2.0));
+
+  for (unsigned int i =0; i < _n4.size(); ++i)
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: space after `=`

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+
+  // Derivative of the residual component wrt tau
+  Real theta, Delta, Psi, dDelta_dt, dPsi_dt;
+  Real dphirdt = 0.0;
+  for (unsigned int i = 0; i < _n1.size(); ++i)
+    dphirdt += _n1[i] * _t1[i] * std::pow(delta, _d1[i]) * std::pow(tau, _t1[i] - 1.0);
+
+  for (unsigned int i = 0; i < _n2.size(); ++i)
+    dphirdt += _n2[i] * _t2[i] * std::pow(delta, _d2[i]) * std::pow(tau, _t2[i] - 1.0) *
+      std::exp( - std::pow(delta, _c2[i]));
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: No space after `(`

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+  for (unsigned int i = 0; i < _n2.size(); ++i)
+    d2phirdt2 += _n2[i] * _t2[i] * (_t2[i] - 1.0) * std::pow(delta, _d2[i]) * std::exp(- std::pow(delta, _c2[i]))
+       * std::pow(tau, _t2[i] - 2.0);
+
+  for (unsigned int i = 0; i < _n3.size(); ++i)
+    d2phirdt2 += _n3[i] * std::pow(delta, _d3[i]) * std::pow(tau, _t3[i]) * std::exp(- _alpha3[i]
+      * std::pow(delta - _eps3[i], 2.0) - _beta3[i] * std::pow(tau - _gamma3[i], 2.0)) * (std::pow(_t3[i] / tau
+      - 2.0 * _beta3[i] * (tau - _gamma3[i]), 2.0) - _t3[i] / tau / tau - 2.0 * _beta3[i]);
+
+ for (unsigned int i = 0; i < _n4.size(); ++i)
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: Indent by 2 spaces

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+  if (temperature > _triple_point_temperature && temperature < _critical_temperature)
+  {
+    Real gas_density = saturatedVapourDensity(temperature);
+    Real liquid_density = saturatedLiquidDensity(temperature);
+
+    if (density < gas_density || density > liquid_density)
+      pressure = pressureSW(density, temperature);
+    else
+      pressure = vapourPressure(temperature);
+    }
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: This has too much indentation, should be 2 spaces...

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+Real
+CO2FluidProperties::rho(Real pressure, Real temperature) const
+{
+  // Check that the input parameters are within the region of validity
+  if (temperature < 216.0 || temperature > 1100.0 || pressure <= 0.0)
+    mooseError("Parameters out of range in CO2FLuidProperties::rho");
+
+  // Also check that the pressure and temperature are not in the solid phase region
+  if (((temperature > _triple_point_temperature) && (pressure > meltingPressure(temperature)))
+  || ((temperature < _triple_point_temperature) && (pressure > sublimationPressure(temperature))))
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: Either add indentation or move `||` to the previous line

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 7]
```diff
@@ -0,0 +1,719 @@
...
+{
+  // This correlation uses temperature in C
+  Real Tc = temperature - _T_c2k;
+  // The parial volume
+  Real V = 37.51 - 9.585e-2 * Tc + 8.74e-4 * Tc * Tc - 5.044e-7 * Tc * Tc * Tc;
+
+  return 1.e6 * _Mco2 / V;
+}
+
+std::vector<Real>
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: `const std::vector<Real> &` will be better unless you plan on changing the constants...

File: modules/fluid_properties/tests/co2/partial_density.i [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    function = tic
+    variable = temperature
+  [../]
+[]
+
+[AuxKernels]
+  [./rhop]
+    type = MaterialRealAux
+     variable = rhop
+     property = partial_density
```
Comment (andrsd) [inline modules/fluid_properties/tests/co2/partial_density.i]: Indent by 2 spaces

File: modules/fluid_properties/tests/co2/sublimation.i [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    family = MONOMIAL
+    order = CONSTANT
+  [../]
+[]
+
+[AuxKernels]
+  [./pmelt]
+    type = MaterialRealAux
+     variable = psub
+     property = sublimation_pressure
```
Comment (andrsd) [inline modules/fluid_properties/tests/co2/sublimation.i]: Indent by 2 spaces

File: modules/fluid_properties/tests/co2/sweos.i [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+  [../]
+  [./cp]
+    type = MaterialRealAux
+     variable = cp
+     property = cp
+  [../]
+  [./c]
+    type = MaterialRealAux
+     variable = c
+     property = c
```
Comment (andrsd) [inline modules/fluid_properties/tests/co2/sweos.i]: Indent by 2 spaces in all these blocks

File: modules/fluid_properties/tests/co2/thermalcond.i [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    function = tic
+    variable = temperature
+  [../]
+[]
+
+[AuxKernels]
+  [./k]
+    type = MaterialRealAux
+     variable = k
+     property = k
```
Comment (andrsd) [inline modules/fluid_properties/tests/co2/thermalcond.i]: Indent by 2 spaces

File: modules/fluid_properties/tests/co2/vapour.i [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+  [../]
+  [./rhovap]
+    type = MaterialRealAux
+     variable = rhovap
+     property = saturated_vapour_density
+  [../]
+  [./rhosat]
+    type = MaterialRealAux
+     variable = rhosat
+     property = saturated_liquid_density
```
Comment (andrsd) [inline modules/fluid_properties/tests/co2/vapour.i]: Indent by 2 spaces in all 3 blocks

File: modules/fluid_properties/include/materials/CO2FluidPropertiesTestMaterial.h [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+  /// Pressure (Pa)
+  const VariableValue & _pressure;
+  /// Temperature (K)
+  const VariableValue & _temperature;
+  /// Sublimation pressure (Pa)
+  MaterialProperty<Real> & _psub;
+  /// Melting pressure (Pa)
+  MaterialProperty<Real> & _pmelt;
+  /// Vapour pressure (Pa)
```
Comment (andrsd) [inline modules/fluid_properties/include/materials/CO2FluidPropertiesTestMaterial.h]: Let's go with the AE spelling: vapor. I am putting the comment only here, but it is on several other lines below

**Review by andrsd — CHANGES_REQUESTED** [review]

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C (hunk 6 above)
Comment (cpgr) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: I think I'll put the `||` on the previous line

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C (hunk 7 above)
Comment (cpgr) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: I'll make this `const`

File: modules/fluid_properties/include/materials/CO2FluidPropertiesTestMaterial.h (hunk 13 above)
Comment (cpgr) [inline modules/fluid_properties/include/materials/CO2FluidPropertiesTestMaterial.h]: I'll change these to the American spelling, and also the function with vapour in its name as well.

Comment (andrsd) [inline modules/fluid_properties/include/materials/CO2FluidPropertiesTestMaterial.h]: Yes, that's what I had in mind, not only variable names, but also methods, comments, etc... 👍

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C (hunk 7 above)
Comment (cpgr) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: If I change this return type to `const std::vector<Real> &` I get a `warning: returning reference to local temporary object`. Is `const std::vector<Real>` ok instead?

Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: You would have to do something like:
```
static std::vector<Real> hc{-8.55445, 4.01195, 9.52345};
return hc;
```
Or you could make the constants a static member of this class and return that.

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+  for (unsigned int i = 0; i < _n2.size(); ++i)
+    dphirdt += _n2[i] * _t2[i] * std::pow(delta, _d2[i]) * std::pow(tau, _t2[i] - 1.0) *
+      std::exp(- std::pow(delta, _c2[i]));
+
+  for (unsigned int i = 0; i < _n3.size(); ++i)
+    dphirdt += _n3[i] * std::pow(delta, _d3[i]) * std::pow(tau, _t3[i]) * std::exp(- _alpha3[i]
+      * std::pow(delta - _eps3[i], 2.0) - _beta3[i] * std::pow(tau - _gamma3[i], 2.0)) * (_t3[i] / tau
+      - 2.0 * _beta3[i] * (tau - _gamma3[i]));
+
+ for (unsigned int i = 0; i < _n4.size(); ++i)
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: Looks like we missed this one: indent by 2 space...

**Review by andrsd — CHANGES_REQUESTED** [review]

File: modules/fluid_properties/src/userobjects/CO2FluidProperties.C (hunk 14 above)
Comment (cpgr) [inline modules/fluid_properties/src/userobjects/CO2FluidProperties.C]: I've fixed this one, and also added a `mooseError` in beta as it isn't implemented. Thanks again for looking at these

**Review by andrsd — APPROVED** [review]

**Comment by andrsd** [comment]
Thanks for the contribution @cpgr

**Pull request merged**